### PR TITLE
Fix TypeScript errors related to new @types/node version

### PR DESCRIPTION
**Pull request by Carthanial**
### Description : 
Upon updating Node from v10 to v12 the previous `@types/node` version became incompatible and upon updating caused errors in these files.

>types.ts

-In Node.js 10, the first argument to querystring.stringify was defined as:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/47bbc5e136c92275fbb51b6458e5b48c06739f81/types/node/v10/querystring.d.ts#L13

In Node.js 12, the type is more strict:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/9ea38ba16ce74f125367446a1697c67b90e8ca42/types/node/v12/querystring.d.ts#L14

This interface conflicts in Node.js 12 because Array<string | number | boolean | undefined> is more loose compared to string[] | number[] | boolean[] 

It's complaining about this issue in multiple files which this change fixes.

>public.api.md

- The change in this file is identical to the change in `types.ts` however there seems to be an issue with eslint formatting that is changing the whole files format upon saving the single line change. This does not break anything but we may want to look into why this is happening.

>plugins_discovery.ts

-  The problem here is the pathStat type was of type unknown in the mergeMap callback and calling `isDirectory()` on it would cause an error.

>file_appender.ts, kbn_url_storage.ts, sha256.ts

- The errors in these files have to do with a change in the way TypeScript handles overloading function params that changed in this version which caused errors of `Argument of type <type> is not assignable to parameter of type <type> Type undefined is not assignable to type <type>.`

>proc.ts

- Changes in this file came from my research in the links provided in issue #48 on the elastic team's process in this same upgrade. I have not found any comments there related to this fix and am not 100% sure on the cause.

#### Research on these errors was performed using the links provided in issue #48

### CODE REVIEW

File: src/core/public/http/types.ts [hunk 1]
```diff
@@ -205,12 +205,7 @@ export interface HttpRequestInit {
 
 /** @public */
 export interface HttpFetchQuery {
-  [key: string]:
-    | string
-    | number
-    | boolean
-    | undefined
-    | Array<string | number | boolean | undefined>;
+  [key: string]: string | number | boolean | string[] | number[] | boolean[] | undefined | null;
```
Comment (Tengda-He) [inline src/core/public/http/types.ts:208]: nit: can we not change the format, put it one line is hard to debug.

**Review by Tengda-He — COMMENTED** [review]
Good detail descriptions!

File: src/core/public/http/types.ts (hunk 1 above)
Comment (Carthanial) [reply src/core/public/http/types.ts:208]: Yeah I can make that change right now

**Comment by Carthanial** [comment]
Noticed an issue @Tengda-He pointed out with formatting that will fix the eslint problem on `public.api.md` . Will fix before merging PR.

**Review by zhyuanqi — APPROVED** [review]

**Pull request merged**